File: lib/transcript/parser.ts
export interface CourseRow {
  term: string;
  subject: string;
  number: string;
  title: string;
  credits: number;
  grade: string | null;
  confidence: number;
}

export interface ParseMetadata {
  unknownLines: number;
  totalLines: number;
  coursesFound: number;
  termsFound: string[];
}

export interface ParseResult {
  courses: CourseRow[];
  metadata: ParseMetadata;
}

const VALID_GRADES = new Set([
  'A+',
  'A',
  'A-',
  'B+',
  'B',
  'B-',
  'C+',
  'C',
  'C-',
  'D+',
  'D',
  'D-',
  'F',
  'CR',
  'NC',
  'P',
  'I',
  'W',
  'T',
]);

const TERM_REGEX = /(Fall|Winter|Spring|Summer)\s+(Semester|Term)\s+20\d{2}/i;

const SUBJECT_PATTERN = '[A-Z]{2,8}(?:\\s+[A-Z]{1,3})?';
const GRADE_PATTERN = '(?:[A-Z][+\\-]?|CR|NC|P|I|W|T)';

const COURSE_REGEX = new RegExp(
  `^\\s*(?<subject>${SUBJECT_PATTERN})\\s+(?<number>\\d{3}[A-Z]?)\\s+(?<section>\\d{3}[A-Z]?)\\s+(?<title>.+?)\\s+(?<credits>\\d+(?:\\.\\d{1,2})?)(?:\\s+(?<grade>${GRADE_PATTERN}))?\\s*$`
);

const COURSE_NO_SECTION_REGEX = new RegExp(
  `^\\s*(?<subject>${SUBJECT_PATTERN})\\s+(?<number>\\d{3}[A-Z]?)\\s+(?<title>.+?)\\s+(?<credits>\\d+(?:\\.\\d{1,2})?)(?:\\s+(?<grade>${GRADE_PATTERN}))?\\s*$`
);

const TRANSFER_REGEX = new RegExp(
  `^\\s*(?<yrtrm>\\d{5})\\s+(?<subject>${SUBJECT_PATTERN})\\s+(?<number>\\d{3}[A-Z]?)\\s+(?<title>.+?)\\s+(?<credits>\\d+(?:\\.\\d{1,2})?)\\s+(?<grade>${GRADE_PATTERN})`,
  'i'
);

const CONTINUATION_GUARD_REGEX = /\d+\.\d{1,2}\s+[A-Z]{1,3}$/;

const SKIP_PATTERNS: RegExp[] = [
  /SEM\s+HR\s+ERN/i,
  /HR\s+GRD/i,
  /GPA\s+\d/i,
  /CUMULATIVE/i,
  /TOTAL\s+CREDITS/i,
  /TEACH\s+CRS\s+SEC/i,
  /AREA\s+NO\./i,
  /YRTRM\s+COURSE/i,
];

const TITLE_WHITESPACE_REGEX = /\s{2,}/g;

const TITLE_CASE_REGEX = /\b([A-Za-z][^\s]*)/g;

function normalizeWhitespace(value: string): string {
  return value.replace(TITLE_WHITESPACE_REGEX, ' ').trim();
}

function toTitleCase(value: string): string {
  return normalizeWhitespace(value).toLowerCase().replace(TITLE_CASE_REGEX, (match) => {
    return match.charAt(0).toUpperCase() + match.slice(1);
  });
}

export function yrtrmToTerm(code: string): string {
  if (code.length !== 5) {
    return 'Unknown';
  }

  const year = code.slice(0, 4);
  const termDigit = code.slice(4);

  const termNameMap: Record<string, string> = {
    '1': 'Winter',
    '3': 'Spring',
    '5': 'Summer',
    '9': 'Fall',
  };

  const termName = termNameMap[termDigit] ?? 'Unknown';
  return `${termName} Term ${year}`;
}

function shouldSkipLine(line: string): boolean {
  return SKIP_PATTERNS.some((pattern) => pattern.test(line));
}

function parseCourseFromMatch(
  groups: Record<string, string>,
  term: string,
  confidence: number
): CourseRow {
  const subject = normalizeWhitespace(groups.subject).toUpperCase();
  const number = groups.number.toUpperCase();

  const credits = Number(groups.credits);
  const gradeRaw = groups.grade ? groups.grade.toUpperCase() : null;
  const grade = gradeRaw && VALID_GRADES.has(gradeRaw) ? gradeRaw : null;

  const title = toTitleCase(groups.title);

  let adjustedConfidence = confidence;
  if (gradeRaw && !grade) {
    adjustedConfidence = Math.min(confidence, 0.75);
  }

  return {
    term,
    subject,
    number,
    title,
    credits,
    grade,
    confidence: adjustedConfidence,
  };
}

export function parseTranscriptText(text: string): ParseResult {
  const lines = text
    .split(/\r?\n/)
    .map((line) => line.replace(/\u00a0/g, ' ').trimEnd())
    .filter((line) => line.trim().length > 0);

  const courses: CourseRow[] = [];
  let unknownLines = 0;
  let currentTerm = 'Unknown';
  let inTransfer = false;
  const termsFound = new Set<string>();

  for (const line of lines) {
    if (shouldSkipLine(line)) {
      continue;
    }

    const termMatch = line.match(TERM_REGEX);
    if (termMatch) {
      currentTerm = termMatch[0]
        .toLowerCase()
        .replace(/\b\w/g, (char) => char.toUpperCase());
      termsFound.add(currentTerm);
      inTransfer = false;
      continue;
    }

    const upperLine = line.toUpperCase();
    if (upperLine.includes('TRANSFER CREDITS RECEIVED') || upperLine.includes('ADVANCED PLACEMENT')) {
      inTransfer = true;
      continue;
    }

    if (upperLine.includes('CURRENT ENROLLMENT')) {
      inTransfer = false;
      continue;
    }

    if (inTransfer && (line.includes('Attended from') || line.includes('Univ'))) {
      continue;
    }

    if (inTransfer) {
      const match = TRANSFER_REGEX.exec(line);
      if (match?.groups) {
        const term = yrtrmToTerm(match.groups.yrtrm);
        const course = parseCourseFromMatch(match.groups, term, 0.75);
        courses.push(course);
        continue;
      }
    }

    const courseMatch = COURSE_REGEX.exec(line);
    if (courseMatch?.groups) {
      const course = parseCourseFromMatch(courseMatch.groups, currentTerm, 0.9);
      courses.push(course);
      continue;
    }

    const altMatch = COURSE_NO_SECTION_REGEX.exec(line);
    if (altMatch?.groups) {
      const course = parseCourseFromMatch(altMatch.groups, currentTerm, 0.85);
      courses.push(course);
      continue;
    }

    const trimmed = line.trim();
    if (
      courses.length > 0 &&
      trimmed.length > 2 &&
      !/^\d/.test(trimmed) &&
      !CONTINUATION_GUARD_REGEX.test(trimmed)
    ) {
      const lastCourse = courses[courses.length - 1];
      lastCourse.title = `${lastCourse.title} ${toTitleCase(trimmed)}`.trim();
      lastCourse.confidence = Math.max(0, lastCourse.confidence - 0.05);
      continue;
    }

    if (trimmed.length > 0) {
      unknownLines += 1;
    }
  }

  const metadata: ParseMetadata = {
    unknownLines,
    totalLines: lines.length,
    coursesFound: courses.length,
    termsFound: Array.from(termsFound).sort(),
  };

  return { courses, metadata };
}

export function validateCourse(course: CourseRow): boolean {
  if (!/^[A-Z]{2,8}(?:\s+[A-Z]{1,3})?$/.test(course.subject)) {
    return false;
  }

  if (!/^\d{3}[A-Z]?$/.test(course.number)) {
    return false;
  }

  if (Number.isNaN(course.credits) || course.credits < 0 || course.credits > 10) {
    return false;
  }

  if (course.grade !== null && !VALID_GRADES.has(course.grade)) {
    return false;
  }

  return true;
}

File: transcript-parser/app/parser.py
"""
Robust transcript parser for BYU-style transcripts.
Handles BYU course work, transfer credits, AP credits, and current enrollment.
"""

import re
from dataclasses import dataclass
from typing import List, Tuple

GRADES = {"A+","A","A-","B+","B","B-","C+","C","C-","D+","D","D-","F","CR","NC","P","I","W","T"}

# Term header pattern: "Fall Semester 2020", "Summer Term 2020", etc.
TERM_RX = re.compile(r"(Fall|Winter|Spring|Summer)\s+(Semester|Term)\s+20\d{2}", re.I)

# BYU course row with section number
COURSE_RX = re.compile(r"""
 ^(?P<subject>[A-Z]{2,8})\s+
   (?P<number>\d{3}[A-Z]?)\s+
   (?P<section>\d{3}[A-Z]?)\s+
   (?P<title>.+?)\s+
   (?P<credits>\d+(?:\.\d{1,2})?)
   (?:\s+(?P<grade>[A-Z][\+\-]?|CR|NC|P|I|W|T))?
 \s*$
""", re.X)

# BYU course row without section (used for AP, some transfers)
COURSE_NOSEC_RX = re.compile(r"""
 ^(?P<subject>[A-Z]{2,8})\s+
   (?P<number>\d{3}[A-Z]?)\s+
   (?P<title>.+?)\s+
   (?P<credits>\d+(?:\.\d{1,2})?)
   (?:\s+(?P<grade>[A-Z][\+\-]?|CR|NC|P|I|W|T))?
 \s*$
""", re.X)

# Transfer course row with YRTRM code
TRANSFER_RX = re.compile(r"""
 ^\s*(?P<yrtrm>\d{5})\s+
   (?P<subject>[A-Z]{2,8})\s+
   (?P<number>\d{3}[A-Z]?)\s+
   (?P<title>.+?)\s+
   (?P<credits>\d+(?:\.\d{1,2})?)\s+
   (?P<grade>[A-Z][\+\-]?|CR|NC|P|I|W|T)
""", re.X)

# Skip lines that are summary/totals
SKIP_PATTERNS = [
    re.compile(r"SEM\s+HR\s+ERN", re.I),
    re.compile(r"HR\s+GRD", re.I),
    re.compile(r"GPA\s+\d", re.I),
    re.compile(r"CUMULATIVE", re.I),
    re.compile(r"TOTAL\s+CREDITS", re.I),
    re.compile(r"TEACH\s+CRS\s+SEC", re.I),  # Header row
    re.compile(r"AREA\s+NO\.", re.I),  # Header row
    re.compile(r"YRTRM\s+COURSE", re.I),  # Transfer header
]


@dataclass
class CourseRow:
    """Represents a parsed course from a transcript."""
    term: str
    subject: str
    number: str
    title: str
    credits: float
    grade: str | None
    confidence: float


def yrtrm_to_term(code: str) -> str:
    """
    Convert BYU YRTRM code to readable term.
    Format: YYYYT where T is term digit
    1 = Winter, 5 = Summer, 9 = Fall, 3 = Spring (if used)
    """
    if len(code) != 5:
        return "Unknown"

    year = code[:4]
    term_digit = code[4]

    term_map = {
        "1": "Winter",
        "3": "Spring",
        "5": "Summer",
        "9": "Fall"
    }

    term_name = term_map.get(term_digit, "Unknown")
    return f"{term_name} Term {year}"


def should_skip_line(line: str) -> bool:
    """Check if line should be skipped (headers, summaries, etc.)."""
    for pattern in SKIP_PATTERNS:
        if pattern.search(line):
            return True
    return False


def parse_text(text: str) -> Tuple[List[CourseRow], dict]:
    """
    Parse transcript text into structured course data.

    Returns:
        Tuple of (courses list, metadata dict)
    """
    lines = [l.rstrip() for l in text.splitlines() if l.strip()]
    courses: List[CourseRow] = []
    unknown_lines = 0
    current_term = "Unknown"
    in_transfer = False
    terms_found = set()

    for i, line in enumerate(lines):
        # Skip summary/header lines
        if should_skip_line(line):
            continue

        # Detect term headers
        mterm = TERM_RX.search(line)
        if mterm:
            current_term = mterm.group(0).title()
            terms_found.add(current_term)
            in_transfer = False
            continue

        # Detect transfer credits section
        if "TRANSFER CREDITS RECEIVED" in line or "ADVANCED PLACEMENT" in line.upper():
            in_transfer = True
            continue

        # Detect current enrollment section
        if "CURRENT ENROLLMENT" in line:
            in_transfer = False
            continue

        # Detect institution headers for transfers (skip but stay in transfer mode)
        if in_transfer and ("Attended from" in line or "Univ" in line):
            continue

        # Try parsing as transfer course (with YRTRM)
        if in_transfer:
            mt = TRANSFER_RX.match(line)
            if mt:
                d = mt.groupdict()
                term = yrtrm_to_term(d["yrtrm"])
                grade = d["grade"]
                title = re.sub(r"\s{2,}", " ", d["title"]).strip().title()

                courses.append(CourseRow(
                    term=term,
                    subject=d["subject"],
                    number=d["number"],
                    title=title,
                    credits=float(d["credits"]),
                    grade=grade if grade in GRADES else None,
                    confidence=0.75
                ))
                continue

        # Try parsing as regular BYU course row (with section)
        m = COURSE_RX.match(line)
        if m:
            d = m.groupdict()
            title = re.sub(r"\s{2,}", " ", d["title"]).strip().title()
            grade = d.get("grade")

            # Validate grade
            if grade and grade not in GRADES:
                grade = None
                conf = 0.7
            else:
                conf = 0.9

            courses.append(CourseRow(
                term=current_term,
                subject=d["subject"],
                number=d["number"],
                title=title,
                credits=float(d["credits"]),
                grade=grade,
                confidence=conf
            ))
            continue

        # Try parsing without section (AP credits, some formats)
        m_nosec = COURSE_NOSEC_RX.match(line)
        if m_nosec:
            d = m_nosec.groupdict()
            title = re.sub(r"\s{2,}", " ", d["title"]).strip().title()
            grade = d.get("grade")

            if grade and grade not in GRADES:
                grade = None
                conf = 0.75
            else:
                conf = 0.85

            courses.append(CourseRow(
                term=current_term,
                subject=d["subject"],
                number=d["number"],
                title=title,
                credits=float(d["credits"]),
                grade=grade,
                confidence=conf
            ))
            continue

        # Check if this might be a title continuation line
        if courses and line.strip() and not re.search(r"\d+\.\d{1,2}\s+[A-Z]{1,3}$", line):
            # Likely a continuation of the previous course title
            last_course = courses[-1]
            continuation = line.strip()

            # Only append if it looks like text (not numbers/codes)
            if not re.match(r"^\d", continuation) and len(continuation) > 2:
                last_course.title = (last_course.title + " " + continuation).strip()
                last_course.confidence -= 0.05
                continue

        # Line didn't match any pattern
        if line.strip():
            unknown_lines += 1

    metadata = {
        "unknown_lines": unknown_lines,
        "total_lines": len(lines),
        "courses_found": len(courses),
        "terms_found": sorted(list(terms_found))
    }

    return courses, metadata


def validate_course(course: CourseRow) -> bool:
    """Validate that a parsed course has valid data."""
    # Subject: 2-8 uppercase letters
    if not re.match(r"^[A-Z]{2,8}$", course.subject):
        return False

    # Number: 3 digits optionally followed by a letter
    if not re.match(r"^\d{3}[A-Z]?$", course.number):
        return False

    # Credits: reasonable range
    if not (0 <= course.credits <= 10):
        return False

    # Grade: either None (current enrollment) or in valid set
    if course.grade is not None and course.grade not in GRADES:
        return False

    return True
